`GetTypeFields` should return `ELEMENT_TYPE_BYREF`.

AaronRobinsonMSFT opened the pull request:
Did not handle the `ByReference<T>` special case.

There is some white space changes as well and spacing that was very confusing. That is modified in this PR.

/cc @noahfalk @hoyosjs

AaronRobinsonMSFT committed:
GetTypeField should return ELEMENT_TYPE_BYREF.

Did not handle the ByReference<T> special case.

tommcdon reviewed (approved):
LGTM, and thanks for adding some code cleanup to the change as well Aaron!

davidwrighton commented on src/coreclr/debug/daccess/dacdbiimpl.cpp:
```diff
@@ -7166,32 +7166,38 @@ HRESULT DacDbiInterfaceImpl::GetObjectFields(COR_TYPEID id, ULONG32 celt, COR_FI
+        }
+        else if (fieldHandle.IsByRef())
+        {
+            corField->id = {};
```
I would have expected corField->id.token2 to be set to the pointed at type of the byref basically how arrays are handled.

AaronRobinsonMSFT replied:
Ah. This makes a lot of good sense. I see your point now. I'll update appropriately.

@davidwrighton any concerns with me ignoring the `ByReference<T>` case? Since this is contributing to the issue where we plan on removing the type, I don't think it is worth the effort but can if we think it has value.

AaronRobinsonMSFT committed:
Add details for byref COR_FIELD.

noahfalk commented on src/coreclr/debug/daccess/dacdbiimpl.cpp:
```diff
@@ -7166,32 +7166,45 @@ HRESULT DacDbiInterfaceImpl::GetObjectFields(COR_TYPEID id, ULONG32 celt, COR_FI
+            if (!tgtType.IsNull())
+            {
+                PTR_MethodTable cmt = tgtType.GetMethodTable();
+                corField->id.token2 = (ULONG64)cmt.GetAddr();
```
I don't think we need to do this, but because this is new behavior we get to define what correct behavior looks like. Two parts of that:
1. If someone calls ICorDebugProcess5::GetTypeForTypeId(id), what type do we expect them to get?
2. Based on the answer to (1), what data do we need to be inside COR_TYPEID so that GetTypeForTypeId() can be implemented correctly?

Imagine we had a type like this:
```
ref Foo
{
    double*  Ptr;
    double[] Array;
    double& ByRef;
}
```

My expectation from code review is that calling GetTypeFields() on Foo, and then calling GetTypeForTypeId() on each of the returned ids would give back types representing IntPtr, double[], and whatever_we_define, respectively. We could follow the precedent set by pointer types and also return IntPtr for ByRef types. Alternatively we could define that ByRefs return an ICorDebugType which represents the exact double&. On the one hand that seems more intuitive, but on the downside it doesn't match the pointer behavior and it might expose lurking bugs in ICorDebugType's own handling of ByRef. We could ask the VS team if they have a preference here.

In terms of the token2 value for arrays, that is a historical artifact because arrays used to have two MethodTable pointers in their runtime object layout so we needed to capture both to adequately describe them. These days my understanding is that the first MethodTable for an array is a complete description of the type and setting token2 is just a remnant that never got cleaned up from the old runtime behavior. Ultimately token1 and token2 should be opaque data so we can use them however we wish, but if we aren't going to have GetTypeForTypeId() reading token2 then there is no value in setting it.

noahfalk replied:
Based on the code [here](https://github.com/dotnet/runtime/blob/d603356dfbf09616e55bd99ed840aa20468a9e57/src/coreclr/debug/daccess/dacdbiimpl.cpp#L2619) token2 appears completely ignored right now.

AaronRobinsonMSFT committed:
Remove setting of token2 in all cases. This value is not used in any
scenario and therefore is more confusing than helpful. Add a commment about
the expectations of the returned values.
